### PR TITLE
[`Idefics`] add image_embeddings option in generate-related methods

### PR DESCRIPTION
**Pull request by leot13**
Update Idefics generate-related functions to allow for precomputed image embeddings

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Could you just run `make style`! 🤗

File: src/transformers/models/idefics/modeling_idefics.py [hunk 1]
```diff
@@ -1118,14 +1148,22 @@ def forward(
...
-        image_seq_len, image_hidden_size = image_hidden_states.size(1), image_hidden_states.size(2)
+            image_seq_len, image_hidden_size = image_hidden_states.size(1), image_hidden_states.size(2)
+        elif self.config.use_resampler and perceiver_embeddings is not None:
+            batch_size, num_images, image_seq_len, image_hidden_size = perceiver_embeddings.size()
+            image_hidden_states = perceiver_embeddings
+        elif not self.config.use_resampler and perceiver_embeddings is None:
+            image_seq_len, image_hidden_size = image_hidden_states.size(1), image_hidden_states.size(2)
+        elif not self.config.use_resampler and perceiver_embeddings is not None:
+            raise ValueError("If perceiver_embeddings are passed, use_resampler should be True")
+
```
Comment (VictorSanh) [inline src/transformers/models/idefics/modeling_idefics.py:1166]: for a more readable flow, i would use nested ifs:
```
if self.config.use_resampler:
    if perceiver_embeddings is None:
        do_thing
else:
    if ...
```
otherwise, the part of the logic looks good to me

**Review by VictorSanh — COMMENTED** [review]
i am missing something in the logic: let's say we call generate with pixel_values. the first time it calls the forward, it computes the image hidden states through vision_encoder (and optionally through the perceiver). how are these image hidden states passed to the second call to the forward?

File: src/transformers/models/idefics/modeling_idefics.py [hunk 2]
```diff
@@ -1435,6 +1460,54 @@ def forward(
...
+        if self.config.use_resampler:
+            if model_kwargs["perceiver_embeddings"] is None:
+                model_kwargs["perceiver_embeddings"] = self.model.perceiver_resampler(image_encoder_embeddings)
+
+            image_seq_len, image_hidden_size = model_kwargs["perceiver_embeddings"].size(1), model_kwargs[
+                "perceiver_embeddings"
+            ].size(2)
+            model_kwargs["perceiver_embeddings"] = model_kwargs["perceiver_embeddings"].view(
+                batch_size, num_images, image_seq_len, image_hidden_size
+            )
```
Comment (VictorSanh) [inline src/transformers/models/idefics/modeling_idefics.py:1488]: do the resizes before assigning it into `model_kwargs["perceiver_embeddings"]`

**Review by VictorSanh — COMMENTED** [review]
the logic looks about right to me.

I doubt that if `return_dict_in_generate` is True, the whole thing works. In particular, this call `model_kwargs["encoder_outputs"]` (in `greedy_search`) will crash.

Also, I am not an expert in the generate function. Computing the encoder hidden states (i.e. the vision/perceiverd hidden states in our case) inside the `prepare...` function seems curious to me. Looking at `modeling_t5.py`, the `encoder_hidden_states` are computed inside the forward of the model (and then returned through `Seq2SeqModelOutput`.

But perhaps it's ok? I will let transformers folks comment instead.

File: src/transformers/models/idefics/modeling_idefics.py (hunk 2 above)
Comment (leot13) [reply src/transformers/models/idefics/modeling_idefics.py:1488]: Sure I'll change that

**Comment by leot13** [comment]
>  Computing the encoder hidden states (i.e. the vision/perceiver hidden states in our case) inside the prepare... function seems curious to me.

Yes I'm not sure it is very standard. 
Maybe a more standard way to do it would be to do an encoder_decoder setting with 'encoder_outputs'. 
That would require adding another input 'encoder_outputs' on top of the image_encoder_embeddings/perceiver_embeddings + a get_encoder() function that would get the encoder and compute the embeddings. 
I think it would add a lot of unnecessary logic, so using _prepare_inputs() to prepare the inputs made more sense to me

File: src/transformers/models/idefics/modeling_idefics.py [hunk 3]
```diff
@@ -1103,11 +1112,16 @@ def forward(
...
-
-        elif pixel_values is not None and image_embeddings is not None:
-            raise ValueError("You cannot specify both pixel_values and image_embeddings at the same time")
+        if (
+            sum(
+                vision_input is not None
+                for vision_input in (pixel_values, image_encoder_embeddings, perceiver_embeddings)
+            )
+            != 1
+        ):
```
Comment (ArthurZucker) [inline src/transformers/models/idefics/modeling_idefics.py:1121]: Maybe just 
```suggestion
        if None not in (pixel_values, image_encoder_embeddings, perceiver_embeddings):
```

File: src/transformers/models/idefics/modeling_idefics.py [hunk 4]
```diff
@@ -1435,6 +1461,98 @@ def forward(
...
+                image_seq_len, image_hidden_size = perceiver_embeddings.size(1), perceiver_embeddings.size(2)
+                model_kwargs["perceiver_embeddings"] = perceiver_embeddings.view(
+                    batch_size, num_images, image_seq_len, image_hidden_size
+                )
+            else:
+                model_kwargs["perceiver_embeddings"] = perceiver_embeddings
+        else:
+            image_seq_len, image_hidden_size = image_encoder_embeddings.size(1), image_encoder_embeddings.size(2)
+            model_kwargs["image_encoder_embeddings"] = image_encoder_embeddings.view(
+                batch_size, num_images, image_seq_len, image_hidden_size
```
Comment (ArthurZucker) [inline src/transformers/models/idefics/modeling_idefics.py:1550]: Regarding @VictorSanh's comment, it makes more sense indeed to compute these values in the forward, and return them as outputs. Our API usually works this way so let's keep this. It should also simplify this function!

File: src/transformers/models/idefics/modeling_idefics.py [hunk 5]
```diff
@@ -106,16 +115,15 @@ def update_model_kwargs_for_generation(outputs, model_kwargs, is_encoder_decoder
...
-            )
-        if "image_attention_mask" in model_kwargs:
-            image_attention_mask = model_kwargs["image_attention_mask"]
-            last_mask = image_attention_mask[:, -1, :].unsqueeze(1)
-            model_kwargs["image_attention_mask"] = last_mask
+    if "attention_mask" in model_kwargs:
+        attention_mask = model_kwargs["attention_mask"]
+        model_kwargs["attention_mask"] = torch.cat(
+            [attention_mask, attention_mask.new_ones((attention_mask.shape[0], 1))], dim=-1
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/idefics/modeling_idefics.py:122]: should the attention mask also default to this outside the generate function? (meaning should this not be done in the forward of the model?)

File: src/transformers/models/idefics/modeling_idefics.py [hunk 6]
```diff
@@ -65,28 +66,36 @@ def expand_inputs_for_generation(
...
-        encoder_outputs["last_hidden_state"] = encoder_outputs.last_hidden_state.index_select(
-            0, expanded_return_idx.to(encoder_outputs.last_hidden_state.device)
+    if model_kwargs["image_encoder_embeddings"] is not None:
+        model_kwargs["image_encoder_embeddings"] = model_kwargs["image_encoder_embeddings"].index_select(
+            0, expanded_return_idx
+        )
+
+    elif model_kwargs["perceiver_embeddings"] is not None:
+        model_kwargs["perceiver_embeddings"] = model_kwargs["perceiver_embeddings"].index_select(
+            0, expanded_return_idx
```
Comment (ArthurZucker) [inline src/transformers/models/idefics/modeling_idefics.py:92]: Is this something that only happens with `generate`? (if yes, then no worries let's keep it here!)

**Review by ArthurZucker — COMMENTED** [review]
Thanks for working on this! Left a few commets. Could you also add some tests to make sure you can pass both expected embeddings and have the expected behaviour! 🤗

File: src/transformers/models/idefics/modeling_idefics.py (hunk 3 above)
Comment (leot13) [reply src/transformers/models/idefics/modeling_idefics.py:1121]: No because I need 2 of them to be None and one to be filled

File: src/transformers/models/idefics/modeling_idefics.py (hunk 6 above)
Comment (leot13) [reply src/transformers/models/idefics/modeling_idefics.py:92]: Not sure what you mean here.. Basically, to the best of my understanding, this is used in beam_search as you need to expand the inputs bsz to match the number of beams

File: src/transformers/models/idefics/modeling_idefics.py (hunk 5 above)
Comment (leot13) [reply src/transformers/models/idefics/modeling_idefics.py:122]: Here, it's just updating the attention_mask to have one more element at each generation

File: src/transformers/models/idefics/modeling_idefics.py (hunk 4 above)
Comment (leot13) [reply src/transformers/models/idefics/modeling_idefics.py:1550]: Okk will look into this

File: src/transformers/models/idefics/modeling_idefics.py (hunk 6 above)
Comment (ArthurZucker) [reply src/transformers/models/idefics/modeling_idefics.py:92]: Ok 👍🏻 then this is all good 😉

Comment (ArthurZucker) [reply src/transformers/models/idefics/modeling_idefics.py:92]: Actually you if you put outputs in `encoder_outputs` the default `https://github.com/ArthurZucker/transformers/blob/main/src/transformers/generation/utils.py#L721` should work!

**Review by VictorSanh — APPROVED** [review]
yaaay! congrats on your first PR to Transformers (I think?)!
lgtm! I'll let @ArthurZucker double confirm

File: src/transformers/models/idefics/modeling_idefics.py [hunk 7]
```diff
@@ -52,6 +53,96 @@
...
+            one for the output of each layer) of shape `(batch_size, sequence_length, hidden_size)`.
+
+            Hidden-states of the model at the output of each layer plus the initial embedding outputs.
+        attentions (:
+            obj:*tuple(torch.FloatTensor)*, *optional*, returned when `output_attentions=True` is passed or when
+            `config.output_attentions=True`): Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size,
+            num_heads, sequence_length, sequence_length)`.
+
+            Attentions weights after the attention softmax, used to compute the weighted average in the self-attention
+            heads.
```
Comment (ArthurZucker) [inline src/transformers/models/idefics/modeling_idefics.py:130]: There is something wrong with the formatting here!

File: src/transformers/models/idefics/modeling_idefics.py [hunk 8]
```diff
@@ -52,6 +53,96 @@
...
+        image_hidden_states (`tuple(torch.FloatTensor)`, *optional*):
+            Tuple of *torch.FloatTensor* (one for the output of the image embeddings, *(batch_size, num_images,
+            sequence_length, hidden_size)*.
+
```
Comment (ArthurZucker) [inline src/transformers/models/idefics/modeling_idefics.py:134]:
```suggestion
            Tuple of `torch.FloatTensor` (one for the output of the image embeddings, `(batch_size, num_images,
            sequence_length, hidden_size)`.

```
should be updated everywhere you use `*` instead of `

File: src/transformers/models/idefics/modeling_idefics.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/models/idefics/modeling_idefics.py:1219]: You can also better format the Error to say which values where not None!

File: src/transformers/models/idefics/modeling_idefics.py [hunk 9]
```diff
@@ -1118,14 +1230,24 @@ def forward(
...
+                image_hidden_states = self.perceiver_resampler(image_hidden_states)
+                image_seq_len, image_hidden_size = image_hidden_states.size(1), image_hidden_states.size(2)
+            else:
+                batch_size, num_images, image_seq_len, image_hidden_size = perceiver_embeddings.size()
+                image_hidden_states = perceiver_embeddings
+        else:
+            if perceiver_embeddings is None:
+                image_seq_len, image_hidden_size = image_hidden_states.size(1), image_hidden_states.size(2)
+            else:
+                raise ValueError("If perceiver_embeddings are passed, use_resampler should be True")
```
Comment (ArthurZucker) [inline src/transformers/models/idefics/modeling_idefics.py:1249]:
```suggestion
                    if self.config.use_resampler:
                        if perceiver_embeddings is None:
                            perceiver_embeddings = self.perceiver_resampler(image_hidden_states)
                        batch_size, num_images, image_seq_len, image_hidden_size = perceiver_embeddings.size()
                        image_hidden_states = perceiver_embeddings
                    elif perceiver_embeddings is None:
                        image_seq_len, image_hidden_size = image_hidden_states.size(1), image_hidden_states.size(2)
                    else:
                        raise ValueError("If `perceiver_embeddings` are passed, use_resampler should be True")
```
I might have messed up the indentation!

**Review by ArthurZucker — APPROVED** [review]
Left a few small nits, good to go otherwise ! Congrats 🔥

File: src/transformers/models/idefics/modeling_idefics.py (hunk 9 above)
Comment (leot13) [reply src/transformers/models/idefics/modeling_idefics.py]: I used this but had to make a few modifications for it to work

File: src/transformers/models/idefics/modeling_idefics.py [hunk 10]
```diff
@@ -1118,14 +1221,23 @@ def forward(
             # Get sequence from the vision encoder
             image_hidden_states = self.vision_model(pixel_values=pixel_values).last_hidden_state
 
-        elif image_embeddings is not None:
-            batch_size, num_images, image_seq_len, image_hidden_size = image_embeddings.size()
-            image_hidden_states = image_embeddings.to(dtype=self.dtype, device=input_ids.device)
+        elif image_encoder_embeddings is not None:
+            batch_size, num_images, image_seq_len, image_hidden_size = image_encoder_embeddings.size()
+            image_hidden_states = image_encoder_embeddings.to(dtype=self.dtype, device=input_ids.device)
```
Comment (ArthurZucker) [inline src/transformers/models/idefics/modeling_idefics.py:1226]: Last nit, is this required? Would think that accelerate handles this since it's not a tensor created on the fly. (unless the casting is what's required not necessarly moving to a different device!

**Review by ArthurZucker — APPROVED** [review]

**Comment by ArthurZucker** [comment]
And if it’s not too much work, a test calling generate to make sure this all works well 🤗

**Comment by leot13** [comment]
I think there's a test here` tests/models/idefics/test_modeling_idefics.py`. Do you have something else in mind?

File: src/transformers/models/idefics/modeling_idefics.py (hunk 10 above)
Comment (leot13) [reply src/transformers/models/idefics/modeling_idefics.py:1226]: Not sure, maybe not. I kept it out of caution when modifying this part. @VictorSanh, do you know if there was a particular reason for adding this?

**Comment by ArthurZucker** [comment]
Okay `class IdeficsModelIntegrationTest(TestCasePlus)` already tests it thanks for checking! Feel free to merge
